### PR TITLE
CFE-3937: fallback to getent if builtin user/group info lookup fails

### PR DESCRIPTION
**Pull request by vpodzime**

### CODE REVIEW

**Comment by vpodzime** [comment]
@cf-bottom jenkins, please, and tell me it builds fine for Windows :crossed_fingers:

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7689/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7689/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7689/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7690/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7690/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7690/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7694/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7694/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7694/

**Comment by vpodzime** [comment]
@cf-bottom jenkins, again, please. And this time with exotics! :scream:

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7695/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7695/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7695/

**Review by olehermanse — APPROVED** [review]
LGTM! Maybe you could add some unit tests for `GetUserGroupInfoFromGetent()`? (string parsing is easy to get wrong)

**Review by larsewi — DISMISSED** [review]
nitpicks

**Comment by vpodzime** [comment]
>  LGTM! Maybe you could add some unit tests for GetUserGroupInfoFromGetent()? (string parsing is easy to get wrong)
Yeah, that would be nice, but making sure it doesn't require `/bin/getent` would be quite messy (`LD_PRELOAD`, `chroot`,...?)

File: libpromises/unix.c [hunk 1]
```diff
@@ -262,4 +240,280 @@ bool ShellCommandReturnsZero(const char *command, ShellType shell)
...
+        if (GetUserGroupInfoFromGetent("passwd", uid_str,
+                                       user_name_buf, buf_size, NULL,
+                                       error_log_level))
+        {
+            /* Found by getent. */
+            return true;
+        }
+        else
+        {
+            Log(error_log_level, "Could not get user name for uid %ju, (getpwuid: %s)",
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/unix.c:367]: We don't put quotes around numbers.

File: libpromises/unix.c [hunk 2]
```diff
@@ -262,4 +240,280 @@ bool ShellCommandReturnsZero(const char *command, ShellType shell)
...
+                                       group_name_buf, buf_size, NULL,
+                                       error_log_level))
+        {
+            /* Found by getent. */
+            return true;
+        }
+        else
+        {
+            Log(error_log_level, "Could not get group name for gid %ju, (getgrgid: %s)",
+                (uintmax_t) gid, (ret == 0) ? "not found" : GetErrorStrFromCode(ret));
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/unix.c:424]: I believe this is quite clear. :)

File: libpromises/unix.c [hunk 3]
```diff
@@ -262,4 +240,280 @@ bool ShellCommandReturnsZero(const char *command, ShellType shell)
...
+bool GetGroupName(gid_t gid, char *group_name_buf, size_t buf_size, LogLevel error_log_level)
+{
+    char buf[GETGR_R_SIZE_MAX] = {0};
+    struct group grp;
+    struct group *result;
+
+    int ret = getgrgid_r(gid, &grp, buf, GETGR_R_SIZE_MAX, &result);
+    if (result == NULL)
+    {
+        char gid_str[32];       /* len("%d" % (2**64 - 1)) == 20 */
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/unix.c:410]: It's a common practice to use powers of two.

[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/unix.c:410]: We don't know the max GID.

File: libenv/sysinfo.c [hunk 4]
```diff
@@ -1576,30 +1577,26 @@ static void OSClasses(EvalContext *ctx)
...
-    else
+    char user_name[CF_SMALLBUF];
+    if (GetCurrentUserName(user_name, sizeof(user_name)))
     {
         char vbuff[CF_BUFSIZE];
 
         if (EvalContextClassGet(ctx, NULL, "SUSE"))
         {
-            snprintf(vbuff, CF_BUFSIZE, "/var/spool/cron/tabs/%s", pw->pw_name);
+            snprintf(vbuff, CF_BUFSIZE, "/var/spool/cron/tabs/%s", user_name);
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libenv/sysinfo.c:1587]: I didn't touch this code.

**Comment by vpodzime** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/7717/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7717/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7717/

**Review by larsewi — APPROVED** [review]
ACK

**Pull request merged**